### PR TITLE
SQS Producer

### LINKED ISSUE #156: AWS SQS producer
## Is your feature request related to a problem? Please describe

We need to create an AWS SQS producer so that we don't have to go through SNS to publish to an SQS queue.

## Is this a bug? Please provide steps to reproduce, a failing test etc.

Nope

## Describe the solution

Create an SQS producer in the `trace` pkg

### PR DESCRIPTION
**Pull request by aziule**
## Which problem is this PR solving?

Resolves #156 

## Short description of the changes

* Add a simple SQS producer with built-in tracing capabilities
* Create a message builder to improve the developer experience
* Test (unit + integration) for standard and FIFO queues
* Add to examples

## Notes

* The current implementation does NOT support message attributes, either standard (string, number or binary) or custom, because the SQS consumer does not support it either. It would be a nice addition to Patron to support attributes.
* The current way examples work, e.g.: one example per "component", was not practical at all for this SQS publisher. I decided to broadcast the incoming message to both SNS and SQS, so that we wouldn't have to create a 2nd SQS consumer in a 6th example. This is why you will see that example 4 creates two traces to example 5.

Jaeger result from the examples:
Screenshot 2020-02-26 at 16 11 24

### CODE REVIEW

File: trace/sqs/message.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+		return nil, errors.New("missing required field: message body")
+	}
+
+	if b.input.QueueUrl == nil {
+		return nil, errors.New("missing required field: message queue URL")
+	}
+
+	// Messages with either a group ID or deduplication ID can't have a delay.
+	// These two attributes are only used for FIFO queues, which don't allow for individual message delays.
+	if (b.input.MessageGroupId != nil || b.input.MessageDeduplicationId != nil) && b.input.DelaySeconds != nil {
```
Comment (mantzas) [inline trace/sqs/message.go:50]: Is this not checked by the package? How does the package behave when these are set?

File: trace/sqs/publisher.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+	carrier := sqsHeadersCarrier{}
+	err = span.Tracer().Inject(span.Context(), opentracing.TextMap, &carrier)
+	if err != nil {
+		return "", fmt.Errorf("failed to inject tracing headers: %w", err)
+	}
+
+	msg.injectHeaders(carrier)
+	msg.setMessageAttribute(correlation.HeaderID, correlation.IDFromContext(ctx))
+
```
Comment (mantzas) [inline trace/sqs/publisher.go:59]: Could we also propagate the correlation id used in logging??

**Review by mantzas — CHANGES_REQUESTED** [review]

File: trace/sqs/message.go (hunk 1 above)
Comment (aziule) [reply trace/sqs/message.go:50]: Unfortunately, it does not check it and returns an empty msg ID. So checking it here avoids sending an unnecessary request.

File: trace/sqs/publisher.go (hunk 2 above)
Comment (aziule) [reply trace/sqs/publisher.go:59]: I'm not certain to understand what you mean. Where is this correlation ID? I have checked the other publishers in the `trace` pkg and it seems the above `correlation.HeaderID` is the only one to be injected.

Comment (mantzas) [reply trace/sqs/publisher.go:59]: do you believe i did not see it??? you are right...

**Review by mantzas — APPROVED** [review]

**Review by moukoublen — APPROVED** [review]

**Pull request merged**